iroha/src/p2p.rs

```diff
@@ -0,0 +1,259 @@
+
+        let mut tx_interval =
+            async_std::stream::interval(Duration::from_secs(self.tx_interval_sec as u64));
+        let mut ping_interval =
+            async_std::stream::interval(Duration::from_secs(self.ping_interval_sec as u64));
+
+        loop {
+            let ping_future = ping_interval.next().fuse();
+            let tx_future = tx_interval.next().fuse();
+
+            pin_mut!(ping_future, tx_future);
+
+            select! {
+                    listen = listen_future => unreachable!(),
+                    ping = ping_future => Self::ping_all(self.state.clone()).await?,
+                    tx = tx_future => Self::broadcast_tx(self.state.clone()).await?,
+            };
+        }
+    }
+
+    async fn broadcast_tx(state: State<NodeState>) -> Result<(), String> {
+        //TODO: check if we have pending tx and then broadcast it
+        let tx = Tx {
+            payload: vec![],
+            peer: Peer {
+                listen_address: state.lock().await.listen_address.clone(),
+            },
+            sent_time: current_time(),
+        };
+        Self::broadcast(state, &Message::PendingTx(tx)).await
+    }
+
+    async fn listen_and_reconnect(&self) {
+        loop {
+            if self.listen().await.is_ok() {
+                unreachable!()
+            }
+        }
+    }
+
+    async fn listen(&self) -> Result<(), String> {
+        async fn handle_request(
+            state: State<NodeState>,
+            request: Request,
+        ) -> Result<Response, String> {
+            Node::handle_message(state, request.payload()).await?;
+            Ok(Response::new())
+        };
+
+        async fn handle_connection(
+            state: State<NodeState>,
+            stream: Box<dyn AsyncStream>,
+        ) -> Result<(), String> {
+            Network::handle_message_async(state, stream, handle_request).await
+        };
+
+        let listen_address = self.state.lock().await.listen_address.clone();
+        Network::listen_async(
+            self.state.clone(),
+            listen_address.as_ref(),
+            handle_connection,
+        )
+        .await
+    }
+
+    pub async fn start_and_connect(&self, peer_address: &str) -> Result<(), String> {
+        let peer = Peer {
+            listen_address: peer_address.to_string(),
+        };
+        self.state.lock().await.peers.insert(peer.clone());
+        let message = Message::NewPeer(Peer {
+            listen_address: self.state.lock().await.listen_address.clone(),
+        });
+        Self::send(&message, peer).await?;
+        self.start().await?;
+        Ok(())
+    }
+
+    async fn send(message: &Message, peer: Peer) -> Result<(), String> {
+        let _response = Network::send_request_to(
+            peer.listen_address.as_ref(),
+            Request::new("/".to_string(), message.into()),
+        )?;
+        Ok(())
+    }
+
+    async fn broadcast(state: State<NodeState>, message: &Message) -> Result<(), String> {
```

Comment:
we can replace extracted function and iterative tasks execution with inline `join!`

Reply:
Actually `broadcast` and `send` should be sync, because we use the underlying `iroha_network::Network::send`. I will change it.

Reply:
So sadly we will not be able to use `join!` here

Reply:
sorry, but I didn't understand why they should be sync?

Reply:
@humb1t Seems like you introduced async `send` for iroha_network in the latest PR, then sure, I will use `join!`